starky/src/prover.rs

```diff
@@ -57,27 +64,149 @@ where
+
+    let zeta = challenger.get_extension_challenge::<D>();
+    // To avoid leaking witness data, we want to ensure that our opening locations, `zeta` and
+    // `g * zeta`, are not in our subgroup `H`. It suffices to check `zeta` only, since
+    // `(g * zeta)^n = zeta^n`, where `n` is the order of `g`.
+    let g = F::Extension::primitive_root_of_unity(degree_bits);
+    ensure!(
+        zeta.exp_power_of_2(degree_bits) != F::Extension::ONE,
+        "Opening point is in the subgroup."
+    );
+    let openings = StarkOpeningSet::new(zeta, g, &trace_commitment, &quotient_commitment);
+
+    // TODO: Add permuation checks
+    let initial_merkle_trees = &[&trace_commitment, &quotient_commitment];
     let fri_params = config.fri_params(degree_bits);
 
-    let opening_proof = fri_proof::<F, C, D>(
-        initial_merkle_trees,
-        lde_polynomial_coeffs,
-        lde_polynomial_values,
-        &mut challenger,
-        &fri_params,
+    let opening_proof = timed!(
         timing,
+        "compute openings proof",
+        PolynomialBatch::prove_openings(
+            &S::fri_instance(zeta, g, rate_bits),
+            initial_merkle_trees,
+            &mut challenger,
+            &fri_params,
+            timing,
+        )
     );
 
-    StarkProof {
+    Ok(StarkProof {
         trace_cap,
         openings,
         opening_proof,
-    }
+    })
+}
+
+/// Computes the quotient polynomials `(sum alpha^i C_i(x)) / Z_H(x)` for `alpha` in `alphas`,
+/// where the `C_i`s are the Stark constraints.
+// TODO: This won't work for the Fibonacci example because the constraints wrap around the subgroup.
+// The denominator should be the vanishing polynomial of `H` without its last element.
+fn compute_quotient_polys<F, C, S, const D: usize>(
+    stark: &S,
+    trace_commitment: &PolynomialBatch<F, C, D>,
+    alphas: &[F],
+    degree_bits: usize,
+    rate_bits: usize,
+) -> Vec<PolynomialCoeffs<F>>
+where
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    S: Stark<F, D>,
+    [(); S::COLUMNS]:,
+    [(); S::PUBLIC_INPUTS]:,
+{
+    let degree = 1 << degree_bits;
+    let points = F::two_adic_subgroup(degree_bits + rate_bits);
+
+    // Evaluation of the first Lagrange polynomial on the LDE domain.
+    let lagrange_first = {
+        let mut evals = PolynomialValues::new(vec![F::ZERO; degree]);
+        evals.values[0] = F::ONE;
+        evals.lde(rate_bits)
+    };
+    // Evaluation of the last Lagrange polynomial on the LDE domain.
+    let lagrange_last = {
+        let mut evals = PolynomialValues::new(vec![F::ZERO; degree]);
+        evals.values[degree - 1] = F::ONE;
+        evals.lde(rate_bits)
+    };
+
+    let z_h_on_coset = ZeroPolyOnCoset::new(degree_bits, rate_bits);
+
+    // Retrieve the LDE values at index `i`.
+    let get_at_index = |comm: &PolynomialBatch<F, C, D>, i: usize| -> [F; S::COLUMNS] {
+        comm.get_lde_values(i).try_into().unwrap()
+    };
+
+    alphas
+        .iter()
```

Comment:
Sorry for the oversight but in hindsight I think I should have given `ConstraintConsumer` multiple alphas, like `alphas: Vec<P::Scalar>`. So we can call `eval_packed_base` only once per point, and have `ConstraintConsumer` apply similar logic as `reduce_with_powers_multi`.

Reply:
(Or can add a TODO for later)

Reply:
> Sorry for the oversight but in hindsight I think I should have given ConstraintConsumer multiple alphas, and an accumulator for each alpha (so alphas and constraint_acc would be Vecs)


Hmm would that be more efficient than using many `ConstraintConsumer`s? It seems like it would just move the loop on `alphas` elsewhere.

Reply:
If I'm reading it right, this currently calls `eval_packed_base` once per `(alpha, x)`, but if we did it the other way it'd be once per `x`, right?

Reply:
Hmm yeah but then we'd have to change `ConstraintConsumer::one()` to something like
```rust
pub fn one(&mut self, constraint: P) {
    for (alpha, acc) in self.alphas.iter().zip(&mut self.constraint_accs) {
        *acc *=alpha;
        *acc += constraint;
    }
}
```
right? So we'd still have to loop on `alphas`, just somewhere else. Either way is fine for me, just want to make sure I understand correctly.